File: scraperlibs/scraperwiki/datastore.php
<?php

class SW_DataStoreClass
{
   private static $m_ds       ;
   protected      $m_socket   ;
   protected      $m_host     ;
   protected      $m_port     ;

   function __construct ($host, $port)
   {
      $this->m_socket    = null     ;
      $this->m_host      = $host    ;
      $this->m_port      = $port    ;
   }


   function connect ()
   {
      /*
      Connect to the data proxy. The data proxy will need to make an Ident call
      back to get the scraperID. Since the data proxy may be on another machine
      and the peer address it sees will have been subject to NAT or masquerading,
      send the UML name and the socket port number in the request.
      */
      if (is_null($this->m_socket))
      {
            $this->m_socket    = socket_create (AF_INET, SOCK_STREAM, SOL_TCP) ;
            socket_connect     ($this->m_socket, $this->m_host, $this->m_port) ;
            socket_getsockname ($this->m_socket, $addr, $port) ;
            $getmsg = sprintf  ("GET /?uml=%s&port=%s HTTP/1.1\n\n", trim(`/bin/hostname`), $port) ;
            socket_send        ($this->m_socket, $getmsg, strlen($getmsg), MSG_EOR) ;
            socket_recv        ($this->m_socket, $buffer, 0xffff, 0) ;
            $result = json_decode ($buffer) ;
            if (! $result[0])
               throw new Exception ($result[1]) ;
      }
   }

   function request($req)
   {
      $this->connect () ;
      $reqmsg  = json_encode ($req) . "\n" ;
      socket_send ($this->m_socket, $reqmsg, strlen($reqmsg), MSG_EOR) ;

      $text = '' ;
      while (true)
      {
            socket_recv ($this->m_socket, $buffer, 0xffff, 0) ;
            if (strlen($buffer) == 0)
               break ;
            $text .= $buffer ;
            if ($text[strlen($text)-1] == "\n")
               break ;
      }

      return json_decode ($text) ;
   }

   function save ($unique_keys, $scraper_data, $date = null, $latlng = null)
   {
       throw new Exception ("This function is no more and shouldn't be accessible") ;
   }


   function postcodeToLatLng ($postcode)
   {
      return $this->request (array ('postcodetolatlng', $postcode)) ;
   }

   function close ()
   {
      socket_send  ($this->m_socket, ".\n", 2, MSG_EOR) ;
      socket_close ($this->m_socket) ;
      $this->m_socket = undef ;
   }

   static function create ($host = null, $port = null)
   {
      if (is_null(self::$m_ds))
         self::$m_ds = new SW_DataStoreClass ($host, $port) ;
      return   self::$m_ds ;
   }
}

?>

File: scraperlibs/scraperwiki/datastore.rb
require 'json'
require 'singleton'
require 'thread'

# the python version of this makes use of a global static copy of the class 
# so the connection is made only once to the dataproxy
# I think the Singleton module implements this magically

class SW_DataStore
    
    @@lock = Mutex.new
    
    include Singleton

    attr_accessor :m_port, :m_host

    def initialize
      @m_socket = nil
      @m_host = nil
      @m_port = nil
    end


    def ensure_connected
        # Connect to the data proxy. The data proxy will need to make an Ident call
        # back to get the scraperID. Since the data proxy may be on another machine
        # and the peer address it sees will have been subject to NAT or masquerading,
        # send the UML name and the socket port number in the request.
        if @m_socket == nil
            @m_socket = TCPSocket.open(@m_host, @m_port)
            proto, port, name, ip = @m_socket.addr()
            getmsg = "GET /?uml=%s&port=%s HTTP/1.1\n\n" % [Socket.gethostname(), port]
            @m_socket.send(getmsg, 0)
            @m_socket.flush()
            buffer = @m_socket.recv(1024)
            result = JSON.parse(buffer)
            if result["status"] != "good"
                raise result["status"]
            end
        end
    end

    def request (req)
        text = ''
        @@lock.synchronize {
          ensure_connected
          reqmsg  = JSON.generate(req) + "\n"

          bytes_sent = 0
          while bytes_sent < reqmsg.length
            bytes_sent += @m_socket.send(reqmsg.slice(bytes_sent, reqmsg.length), 0)
          end
          @m_socket.flush()

          while true
            buffer = @m_socket.recv(1024)
            if buffer.length == 0
                break
            end
            text += buffer
            if text[-1] == "\n"[0]
                break
            end
          end
        }
        return JSON.parse(text)
    end

    def SW_DataStore.create(host = nil, port = nil)
        instance = SW_DataStore.instance
        # so, it might be intended that the host and port are
        # set once, never to be changed, but this is ruby so
        # there's no way to guarantee that.
        if host && port && instance.m_port.nil? && instance.m_host.nil?
          instance.m_host = host
          instance.m_port = port
        elsif host && port
          raise "Can't change host and port once connection made"
        elsif !(instance.m_port) || !(instance.m_host)
          raise "Can't return a datastore without port/host information"
        end
        instance
    end
end
